src/apps/MASSVEC3DPA.hpp

```diff
@@ -0,0 +1,298 @@
+///          for (int dx = 0; dx < MVPA_D1D; ++dx) {
+///            u += X[dz][dy][dx] * B[qx][dx];
+///          }
+///          DDQ[dz][dy][qx] = u;
+///        }
+///      }
+///    }
+///
+///    for (int dz = 0; dz < MVPA_D1D; ++dz) {
+///      for (int qy = 0; qy < MVPA_Q1D; ++qy) {
+///        for (int qx = 0; qx < MVPA_Q1D; ++qx) {
+///
+///          double u = 0.0;
+///          for (int dy = 0; dy < MVPA_D1D; ++dy) {
+///            u += DDQ[dz][dy][qx] * B[qy][dy];
+///          }
+///          DQQ[dz][qy][qx] = u;
+///        }
+///      }
+///    }
+///
+///    for (int qz = 0; qz < MVPA_Q1D; ++qz) {
+///      for (int qy = 0; qy < MVPA_Q1D; ++qy) {
+///        for (int qx = 0; qx < MVPA_Q1D; ++qx) {
+///
+///          double u = 0.0;
+///          for (int dz = 0; dz < MVPA_D1D; ++dz) {
+///            u += DQQ[dz][qy][qx] * B[qz][dz];
+///          }
+///          QQQ[qz][qy][qx] = u * D(qx, qy, qz, e);
+///        }
+///      }
+///    }
+///
+///    for (int qz = 0; qz < MVPA_Q1D; ++qz) {
+///      for (int qy = 0; qy < MVPA_Q1D; ++qy) {
+///        for (int dx = 0; dx < MVPA_D1D; ++dx) {
+///
+///          double u = 0.0;
+///          for (int qx = 0; qx < MVPA_Q1D; ++qx) {
+///            u += QQQ[qz][qy][qx] * Bt[dx][qx];
+///          }
+///          QQD[qz][qy][dx] = u;
+///        }
+///      }
+///    }
+///
+///    for (int qz = 0; qz < MVPA_Q1D; ++qz) {
+///      for (int dy = 0; dy < MVPA_D1D; ++dy) {
+///        for (int dx = 0; dx < MVPA_D1D; ++dx) {
+///
+///          double u = 0.0;
+///          for (int qy = 0; qy < MVPA_Q1D; ++qy) {
+///            u += QQD[qz][qy][dx] * Bt[dy][qy];
+///          }
+///          QDD[qz][dy][dx] = u;
+///        }
+///      }
+///    }
+///
+///    for (int dz = 0; dz < MVPA_D1D; ++dz) {
+///      for (int dy = 0; dy < MVPA_D1D; ++dy) {
+///        for (int dx = 0; dx < MVPA_D1D; ++dx) {
+///
+///          double u = 0.0;
+///          for (int qz = 0; qz < MVPA_Q1D; ++qz) {
+///            u += QDD[qz][dy][dx] * Bt[dz][qz];
+///          }
+///          mvpaY_(dx, dy, dz, c, e) = u;
+///        }
+///      }
+///    }
+///
+///  } // element loop
+///
+
+#ifndef RAJAPerf_Apps_MASSVEC3DPA_HPP
+#define RAJAPerf_Apps_MASSVEC3DPA_HPP
+
+#define MASSVEC3DPA_DATA_SETUP                                                 \
+  Real_ptr B = m_B;                                                            \
+  Real_ptr Bt = m_Bt;                                                          \
+  Real_ptr D = m_D;                                                            \
+  Real_ptr X = m_X;                                                            \
+  Real_ptr Y = m_Y;                                                            \
+  Index_type NE = m_NE;
+
+#include "FEM_MACROS.hpp"
+#include "common/KernelBase.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+// Number of Dofs/Qpts in 1D
+#define MVPA_D1D 3
+#define MVPA_Q1D 4
+#define MVDIM 3
+#define mvpaB_(x, y) B[x + MVPA_Q1D * y]
```

Comment:
Please use MVPA_ at the beginning of all these macros to make it clear that they are from this file and less likely to clash with other macros.

Reply:
updated